src/core/agent/services/keriaNotificationService.ts

```diff
@@ -0,0 +1,743 @@
+          };
+          await this.basicStorage.createOrUpdateBasicRecord(
+            new BasicRecord({
+              id: MiscRecordId.KERIA_NOTIFICATION_MARKER,
+              content: notificationQuery,
+            })
+          );
+          continue;
+        }
+        if (notificationQuery.nextIndex > 0) {
+          // Since the first item is the (next index - 1), we can ignore it
+          notifications.notes.shift();
+        }
+        for (const notif of notifications.notes) {
+          try {
+            await this.processNotification(notif, callback);
+            const nextNotificationIndex = notificationQuery.nextIndex + 1;
+            notificationQuery = {
+              nextIndex: nextNotificationIndex,
+              lastNotificationId: notif.i,
+            };
+            await this.basicStorage.createOrUpdateBasicRecord(
+              new BasicRecord({
+                id: MiscRecordId.KERIA_NOTIFICATION_MARKER,
+                content: notificationQuery,
+              })
+            );
+          } catch (error) {
+            /* eslint-disable no-console */
+            console.error("Error when process a notification", error);
+          }
+        }
+        if (!notifications.notes.length) {
+          await new Promise((rs) =>
+            setTimeout(rs, KeriaNotificationService.POLL_KERIA_INTERVAL)
+          );
+        }
+      }
+    } catch (error) {
+      /* eslint-disable no-console */
+      console.error("Error at pollNotificationsWithCb", error);
+      setTimeout(
+        this.pollNotificationsWithCb,
+        KeriaNotificationService.POLL_KERIA_INTERVAL,
+        callback
+      );
+    }
+  }
+
+  startNotification() {
+    this.loggedIn = true;
+  }
+
+  stopNotification() {
+    this.loggedIn = false;
+  }
+
+  async deleteNotificationRecordById(
+    id: string,
+    route: NotificationRoute
+  ): Promise<void> {
+    if (!/^\/local/.test(route)) {
+      await this.markNotification(id);
+    }
+    await this.notificationStorage.deleteById(id);
+  }
+
+  async processNotification(
+    notif: Notification,
+    callback: (event: KeriaNotification) => void
+  ) {
+    if (notif.r) {
+      return;
+    }
+    let shouldCreateRecord = true;
+    if (notif.a.r === NotificationRoute.ExnIpexApply) {
+      shouldCreateRecord = await this.processExnIpexApplyNotification(notif);
+    } else if (notif.a.r === NotificationRoute.ExnIpexGrant) {
+      shouldCreateRecord = await this.processExnIpexGrantNotification(notif);
+    } else if (notif.a.r === NotificationRoute.MultiSigRpy) {
+      shouldCreateRecord = await this.processMultiSigRpyNotification(notif);
+    } else if (notif.a.r === NotificationRoute.MultiSigIcp) {
+      shouldCreateRecord = await this.processMultiSigIcpNotification(notif);
+    } else if (notif.a.r === NotificationRoute.MultiSigExn) {
+      shouldCreateRecord = await this.processMultiSigExnNotification(notif);
+    } else if (notif.a.r === NotificationRoute.ExnIpexAgree) {
```

Comment:
Small thing but can you move this to the _first_ `else if`. So we have an ordering if IPEX apply, agree, grant together. And then move onto the others